pkg/checker/functools.go

```diff
@@ -0,0 +1,113 @@
+	}
+	return func() error {
+		ret := reflect.ValueOf(f).Call(argValues)
+		// len(ret) is assumed to be 1 and of type `error`, because of type
+		// `CheckRunner`
+		if ret[0].IsNil() {
+			return nil
+		}
+		return ret[0].Interface().(error)
+	}, nil
+}
+
+func makeRemediationRunner(f interface{}, args ...interface{}) (RemediationRunner, error) {
+	if err := validateFunction(f, args...); err != nil {
+		return nil, fmt.Errorf("invalid function: %v", err)
+	}
+	argValues := make([]reflect.Value, 0, len(args))
+	for _, arg := range args {
+		argValues = append(argValues, reflect.ValueOf(arg))
+	}
+	return func() error {
+		ret := reflect.ValueOf(f).Call(argValues)
+		// len(ret) is assumed to be 1 and of type `error`, because of type
+		// `RemediationRunner`
+		if ret[0].IsNil() {
+			return nil
+		}
+		return ret[0].Interface().(error)
+	}, nil
+}
+
+// validateFunction checks whether `f` is a function, and whether
+// the passed parameters have the correct number and types.
+func validateFunction(f interface{}, args ...interface{}) error {
+	typ := reflect.TypeOf(f)
+	errorType := reflect.TypeOf((*error)(nil)).Elem()
+	if kind := typ.Kind(); kind != reflect.Func {
+		return fmt.Errorf("invalid check runner, want %v, got %v", reflect.Func, kind)
+	}
+	// check output parameter type(s). The function must return an error
+	// TODO: assess if we want to also allow (value, error) as acceptable return
+	//       types. This means also changing how values are returned in
+	//       `makeCheckRunner` and `makeRemediationRunner`.
+	switch {
+	case typ.NumOut() == 1:
+		// return type must be error
+		if typ.Out(0) != errorType {
+			return fmt.Errorf("invalid return type, want %v, got %v", errorType, typ.Out(0))
+		}
+	default:
+		return fmt.Errorf("invalid number of returned parameters, want 1, got %d", typ.NumOut())
+	}
+	// check input parameter types
+	if typ.IsVariadic() {
+		if len(args) < typ.NumIn()-1 {
+			return fmt.Errorf("wrong number of parameters for variadic function, want at least %d, got %d", typ.NumIn()-1, len(args))
+		}
+		// check non-variadic arguments
+		for idx := 0; idx < typ.NumIn()-1; idx++ {
+			want, got := typ.In(idx).Kind(), reflect.TypeOf(args[idx]).Kind()
+			if want != got {
+				return fmt.Errorf("wrong non-variadic parameter type at index %d: want %v, got %v", idx, want, got)
+			}
+		}
+		// check variadic arguments
+		vargs := typ.In(typ.NumIn() - 1)
+		if vargs.Kind() != reflect.Slice {
+			return fmt.Errorf("invalid variadic arguments: want %v, got %v", reflect.Slice, vargs.Kind())
+		}
+		for idx := typ.NumIn() - 1; idx < len(args); idx++ {
+			argType := reflect.TypeOf(args[idx])
+			if argType.Kind() == reflect.Interface {
+				if !argType.Implements(vargs.Elem()) {
+					return fmt.Errorf("invalid interface type for variadic argument at index %d: %v does not implement %v", idx, reflect.TypeOf(args[idx]), vargs.Elem().Kind())
+				}
+			} else {
+				if argType.Kind() != vargs.Elem().Kind() {
```

Comment:
Am I understand correctly that we check only `Kind` (which is not enough to guarantee if it will work) because we do not want to handle complex logic with "underlying types"?

Reply:
correct, in this case I only check that, if this is not an interface type, the expected type matches 1:1 the provided type. In which case would this fail?

Reply:
This is not a comparison of types, but comparison of type kinds. For example `[]uint8` and `[]uint16` are of the same kind: `reflect.Slice`.

Reply:
I understand the problem, and yes, I overlooked it. Do you think it's valuable to be stricter here, or the code would become too complicated? I want to find the right balance between complexity and risk (possibly very low) that the call fails at runtime

Reply:
I think here the case where we should improve continuously. What we have now is already something good. If we will face into a problem caused by this softness, then it will be time to add the complexity.

Reply:
After a little-bit of thinking. I think we just should compare types instead of comparing kinds.

Reply:
in other words change this to `if argType != vargs.Elem()`, thus comparing only whether it's a slice rather than a slice of (for example) strings?

Btw I will update the PR with the unit tests so reasoning about this will become easier. Thanks for the feedback so far!

Reply:
> in other words change this to `if argType != vargs.Elem()`

Yes.

> thus comparing only whether it's a slice rather than a slice of (for example) strings?

The opposite. Few examples:
| `argType` represents | `varg.Elem()` represents | `argType == vargs.Elem()` | `argType.Kind() == vargs.Elem().Kind()` |
| --- | --- | --- | --- |
| `[]uint8` | `[]uint8` | `true` | `true` |
| `[]uint8` | `[]uint16` | `false` | `true` |
| `[]uint8` | `uint8` | `false` | `false` |

See also: https://play.golang.org/p/zB3BkhvGWah

Reply:
I thought one thing and typed another, sorry. I meant the opposite as you pointed out. I agree with your advice, change upcoming